### PR TITLE
Dev-docs: Testing

### PR DESCRIPTION
**Pull request by brandonhewer**
This is the dev-docs move of the following Testing pages:
* [Unit Test Good Practice](http://www.mantidproject.org/Unit%20Test%20Good%20Practice)
* [Data and Helpers for unit tests](http://www.mantidproject.org/Data%20and%20Helpers%20for%20unit%20tests) - Renamed to "Testing Utilities"
* [Data Files in Mantid](http://www.mantidproject.org/Data%20Files%20in%20Mantid) - Renamed to "Data Files for Testing"
* [Running the unit tests](http://www.mantidproject.org/Running%20the%20unit%20tests)
* [Writing Performance Tests](http://www.mantidproject.org/Writing%20Performance%20Tests)
* [System Tests](http://www.mantidproject.org/System%20Tests)

**To test:**
Build dev-docs and make sure the new page reads and renders clearly.
Ensure the changes in #22082 have been implemented

Refs #22082

### CODE REVIEW

File: dev-docs/source/UnitTestGoodPractice.rst [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+Again these can often conflict.
+
+-  Each test class should be named after the class it is testing (e.g.
+   tests for the AlgoithmFactory should go in a AlgorithmFactoryTest
+   class).
+-  Each test within a test class should use a descriptive test name,
+   prefixed with test (tests for the CreateAlgorithm method would be
+   included in testCreateAlgorithm). If there are specific tests for
+   failure situations then these should be added to the end (e.g.
+   testCreateAlgorithmNoAlgorithmException). THE AIM IS THAT FROM THE
```
Comment (DanNixon) [inline dev-docs/source/UnitTestGoodPractice.rst]: Use inline code formatting for function name.

File: dev-docs/source/UnitTestGoodPractice.rst [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      static NameOfTest *createSuite() { return new NameOfTest(); }
+      static void destroySuite(NameOfTest *suite) { delete suite; }
+
+   where ``NameOfTest`` is the name of the test class. Without this, the
+   class is turned into a static meaning that the constructor is run at
+   initialisation even if (via an argument) you are not going to run that
+   particular test suite. Also, this can cause problems if running tests in
+   parallel.
+
+-  Be cautious in use of the setUp() and tearDown() methods. Be aware
```
Comment (DanNixon) [inline dev-docs/source/UnitTestGoodPractice.rst]: More code formatting.

File: dev-docs/source/UnitTestGoodPractice.rst [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+or save a file then you should not use a file in your unit tests.
+
+How do I get a workspace filled with data?
+    Firstly you want to think about how much data you really need, unit
+    tests need to be fast so you don't want too much data.
+    Secondly you should use and extend helper classes (like
+    `1 <https://github.com/mantidproject/mantid/blob/master/Code/Mantid/Framework/TestHelpers/inc/MantidTestHelpers/WorkspaceCreationHelper.h>`__)
+    to provide the workspaces for you. Keep things as generic as you can
+    and it will help you and others for other tests.
+    More details of this will be provided at `Testing Utilities <TestingUtilities>`__.
```
Comment (DanNixon) [inline dev-docs/source/UnitTestGoodPractice.rst]: This link does not work for me.

Is this a result of the pages not yet being in the doctree or is there a `.html` missing?

File: dev-docs/source/UnitTestGoodPractice.rst [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+#######
+
+Mocking is a very powerful tool that allows you to simulate components
+in your unit environment and check how your code operates within this
+environment. Mocking allows you to avoid creating Fake objects of any
+kind, and results in fast executing code with a very high test coverage.
+See `Mocking <Mocking>`__ in Mantid to find out what it is and how it
+works.
+
+.. figure:: Mocking.png
```
Comment (DanNixon) [inline dev-docs/source/UnitTestGoodPractice.rst]: This image doesn't render for me.

File: dev-docs/source/DataFilesForTesting.rst [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+This page gives an overview of how data files are managed within Mantid.
+
+
+Motivation
+##########
+
+Some unit tests use a small amount of data that is created by the test
+harness and others load data from a file. Take the example of
+**ApplyCalibrationTest**. In its first test, testSimple, it creates a
```
Comment (DanNixon) [inline dev-docs/source/DataFilesForTesting.rst]: Use code formatting.

File: dev-docs/source/DataFilesForTesting.rst [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+   ln -s ~/Dropbox\ \(ORNL\)/MantidExternalData ~
+
+Proxy Settings
+--------------
+
+If you are sitting behind a proxy server then the shell or Visual studio
+needs to know about the proxy server. You must set the ``http_proxy``
+environment variable to
+```http://HOSTNAME:PORT`` <http://HOSTNAME:PORT>`__.
```
Comment (DanNixon) [inline dev-docs/source/DataFilesForTesting.rst]: This should just be `http://HOSTNAME:PORT` (i.e. no link).

File: dev-docs/source/SystemTests.rst [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+   def requiredMemoryMB(self):
+      return 2000
+
+The above function limits the test to run on a machine where there is at
+least 2GB of free memory.
+
+.. _target-platform-based-on-free-memory-1:
+
+Target Platform Based on Free Memory
```
Comment (DanNixon) [inline dev-docs/source/SystemTests.rst]: Duplicate heading

File: dev-docs/source/SystemTests.rst [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+new file <Data_Files_in_Mantid#Adding_A_New_File>`__ for how to add new
+files.
+
+Jenkins Test Schedule and Output
+################################
+
+They are run periodically as Jenkins jobs. If the tests pass, then the
+installer(s) relating to the version under test are copied to the
+`download server <http://download.mantidproject.org/>`__, where they are
+marked as the development build, and the subversion tag
```
Comment (DanNixon) [inline dev-docs/source/SystemTests.rst]: Out of date information (SVN tags).

File: dev-docs/source/TestingUtilities.rst [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+C++
+---
+
+The following helper files have been found in the
+`Mantid/Framework/TestHelpers <http://github.com/mantidproject/mantid/tree/master/Framework/TestHelpers>`__
+package:
+
+-  `BinaryOperationMDTestHelper <http://doxygen.mantidproject.org/d1/d4f/namespaceBinaryOperationMDTestHelper.html>`__
+-  **ComponentCreationHelper**
```
Comment (DanNixon) [inline dev-docs/source/TestingUtilities.rst]: Extra bold text not needed here.

File: dev-docs/source/TestingUtilities.rst [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+-  `ScopedFileHelper <http://doxygen.mantidproject.org/d7/d7f/classScopedFileHelper_1_1ScopedFile.html#details>`__
+   This creates a file that is automatically deleted when no longer needed.
+-  `WorkspaceCreationHelper <http://doxygen.mantidproject.org/d1/db6/namespaceWorkspaceCreationHelper.html>`__
+   This creates simple workspaces that can be used in a unit test. One of these workspaces has a full instrument.
+
+Python
+------
+
+There are some testhelpers which are only available in Python, they can
+be found in the **testhelpers**-package.
```
Comment (DanNixon) [inline dev-docs/source/TestingUtilities.rst]: Use code formatting for "testhelpers"

File: dev-docs/source/WritingPerformanceTests.rst [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+Running Performance Tests
+#########################
+
+Performance tests targets are not added by default when running
+cmake/make. To add them as ctest targets, enable them with the flag:
+
+.. code-block:: sh
+
+cmake -DCXXTEST_ADD_PERFORMANCE=TRUE
```
Comment (DanNixon) [inline dev-docs/source/WritingPerformanceTests.rst]: Indentation is needed on the examples here, currently they render as empty code blocks.

File: dev-docs/source/WritingPerformanceTests.rst [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+   timings will only relate to the target code you wish to measure.
+
+Jobs that monitor performance
+#############################
+
+There is a job in Jenkins (our continuous integration system) that runs
+the performance test suite and generates output that enables us to
+easily monitor timings. The job runs a set of `performance tests on the
+master branch of
+Mantid <http://builds.mantidproject.org/view/Tests/job/performance_tests_master/Master_branch_performance_tests/>`__.
```
Comment (DanNixon) [inline dev-docs/source/WritingPerformanceTests.rst]: http://builds.mantidproject.org/job/master_performancetests2/ is probably a more relevant link to go here (obviously the text also needs updated to reflect this is a machine at the ESS).

**Review by DanNixon — CHANGES_REQUESTED** [review]
See comments.

**Comment by martyngigg** [comment]
@brandonhewer @DanNixon Is there more work to do here or is this ready to go?

**Comment by DanNixon** [comment]
There are two comments that still need addressing:
- https://github.com/mantidproject/mantid/pull/22150/files#r176696370
- https://github.com/mantidproject/mantid/pull/22150/files#r176696121

**Comment by brandonhewer** [comment]
@martyngigg @DanNixon Duplicate heading and outdated section has been removed.

**Comment by DanNixon** [comment]
I think mentioning the nightly master pipeline is still worthwhile, it should just be updated to reflect the current workflow, or at very least link to the master pipeline view in Jenkins (http://builds.mantidproject.org/view/Master%20Pipeline/) and explain how the nightly build copy works.

**Comment by DanNixon** [comment]
Actually do we have any pages that are specifically about Jenkins? I think what I was suggesting to include would be more appropriate there as it was not particularly specific to the system tests.

**Review by DanNixon — APPROVED** [review]

**Comment by martyngigg** [comment]
[AutomatedBuildProcess](/mantidproject/mantid/blob/master/dev-docs/source/AutomatedBuildProcess.rst) looks like the best place.

**Pull request merged**